105-metrics-reporter-kip-714.md

```diff
@@ -0,0 +1,104 @@
+
+When a client retrieves its metric subscriptions, it is assigned a unique client instance Id (UUID). This Id is used to identify all the metrics for a specific client instance, even if it sends them to different brokers. This Id is always added as a label using the `client_instance_id` name to all metric series.
+
+### Configurations
+
+The `client_instance_id` label is nice to identify all metrics from a client but does not enable to identify which client it is. Fortunately every time clients send metrics, they attach a bunch of metadata to identify themselves. This metadata is an [`AuthorizableRequestContext`](https://kafka.apache.org/40/javadoc/org/apache/kafka/server/authorizer/AuthorizableRequestContext.html) object that contains the client address, client Id, correlation Id, listener name, principal, request type, request version and security protocol from the client.
+
+A few of these fields are of very low value for metrics:
+- request type: This is always `72` which is the [`PushTelemetry`](https://kafka.apache.org/protocol#The_Messages_PushTelemetry) API key.
+- request version: This is the version of the `PushTelemetry` request. As of Kafka 4.0, it is always `0`.
+- correlation Id: This starts at zero for each client and keeps increasing every time the client sends metrics. This shouldn't be used as a label as it would effectively create a new metric series each time.
+
+The reporter can convert the other fields as labels and add them to the metric series. In many cases it does not make sense to add all of them, for example if the cluster is behind a proxy the client address will always be the same. Also using all of them can create high cardinality labels series which can be problematic in Prometheus.
+
+I propose introducing a configuration to select the metadata fields to use as labels: 
+- `prometheus.metrics.reporter.telemetry.labels`: List of label names in client metrics. The valid names are `client_id`, `listener_name`, `security_protocol`, `principal`, `client_address`. This defaults to `client_id`. This configuration is reconfigurable at runtime.
+
+### Client metrics life cycle
+
+When the reporter receives metrics from a client, it converts them and adds them into Prometheus registry, ready to be collected when the `GET /metrics` endpoint is scraped. The subscription specifies an interval for the client to send metrics. However when metrics are received, the interval or any details about the subscription that caused this emission is not available. 
+
+So we need a mechanism to remove metrics from the registry when a client stops emitting metrics, otherwise if a client shuts down, the metrics reporter will keep reporting its metrics. As we don't have the interval, the simplest approach is to delete client metrics whenever they are scraped. While this may create gaps in metric series, this ensures that metrics collected are always valid. 
```

Comment:
What if we have two Prometheus instances scraping from the same SMR endpoint? Depending on configuration, one of them could be starved and never see client metrics.

Reply:
Is it common to have 2 Prometheus servers scraping the same endpoints? I have to admit that's not a scenario I've considered.

Reply:
I mean, why not? You could easily have dedicated Prometheus instances, for example one dedicated to server metrics and another one to client metrics, or maybe one instance collecting a critical set of metrics from all Kafka clusters, and then a smaller instance for each cluster with a bigger set of metrics.

What about deleting removed metrics on telemetry pushes? I don't know if there is any technical limitation, but you already have the old set in the Prometheus registry, so it should be possible to remove metrics by comparing them with the incoming set.

Reply:
> What about deleting removed metrics on telemetry pushes? I don't know if there is any technical limitation, but you already have the old set in the Prometheus registry, so it should be possible to remove metrics by comparing them with the incoming set.

The issue is when the client stops pushing new metrics, you then keep the last value you received and start exposing stale metrics.

Reply:
Yes, that's a problem that we could solve with a configurable timeout, that should be greater than the scrape period, or the largest scrape period in case of multiple Prom instances with different scrape periods. We could set a default value according to the default Prom scrape period.